### PR TITLE
chore: revamp CI v0.8.25

### PR DESCRIPTION
**Pull request by antonbaliasnikov**
# What ❔

* [x] Unify and update CI scripts: introduce matrix jobs, use common integration tests and benchmarks (remove necessity to create special configs)
* [x] Rename `ZKEVM` to `ZKsync` everywhere
* [x] Fix versioning - use only the second part of the tag as the `ZKsync` version
* [x] Make the ZKsync part of the changelog more unified

## Why ❔

Fixes [CPR-1689](https://linear.app/matterlabs/issue/CPR-1689/update-ci-of-era-solidity-repository)

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.

### CODE REVIEW

**Comment by hedgar2017** [comment]
@antonbaliasnikov why the main workflow runs Y+M3B3?

**Comment by antonbaliasnikov** [comment]
> @antonbaliasnikov why the main workflow runs Y+M3B3?

It's in progress, I'm creating cache to speed up runs.

**Comment by hedgar2017** [comment]
> > @antonbaliasnikov why the main workflow runs Y+M3B3?
> 
> It's in progress, I'm creating cache to speed up runs.

Ok.
So the mode will be `0.8.25` in the end, right?

**Comment by antonbaliasnikov** [comment]
> > > @antonbaliasnikov why the main workflow runs Y+M3B3?
> > 
> > 
> > It's in progress, I'm creating cache to speed up runs.
> 
> Ok. So the mode will be `0.8.25` in the end, right?

Yes

**Comment by antonbaliasnikov** [comment]
@hedgar2017 , @abinavpp , hey guys!

The final tests are running now, I'm opening this for a final review. I properly tested it with many versions and aligned CI to properly work on all of them.

@abinavpp , as soon as we merge it, you can rebase from `0.8.25` and unblock testing on your branch.

File: cmake/templates/BuildInfo.h.in
Comment (abinavpp) [inline cmake/templates/BuildInfo.h.in]: NIT: Is there any problem with renaming this to SOL_VERSION_ZKSYNC? To keep the naming consistent with the rest.

File: .github/actions/solc-build/action.yml
Comment (abinavpp) [inline .github/actions/solc-build/action.yml:51]: NIT: Use one of these?

File: cmake/templates/BuildInfo.h.in
Comment (antonbaliasnikov) [reply cmake/templates/BuildInfo.h.in]: In some versions it's `SOL_VERSION_ZKEVM`, and in some, it's `SOLC_VERSION_ZKEVM`, I can change to `SOL_VERSION_ZKSYNC` everywhere, wdyt?

File: .github/actions/solc-build/action.yml
Comment (antonbaliasnikov) [reply .github/actions/solc-build/action.yml:51]: This is mandatory to be able to build reference versions that can be tested and benchmarked correctly, if we remove the old ones now, tests will fail for the reference.

File: cmake/templates/BuildInfo.h.in
Comment (antonbaliasnikov) [reply cmake/templates/BuildInfo.h.in]: Or actually I can change `SOL_VERSION_ZKEVM` to `SOL_VERSION_ZKSYNC` and `SOLC_VERSION_ZKEVM` `SOLC_VERSION_ZKSYNC` accordingly when I'm doing the auto-update.

File: .github/workflows/build-release.yaml
Comment (abinavpp) [inline .github/workflows/build-release.yaml:149]: Is it possible to have one step for building solc? I see a "Build solc" step in action.yml. If not, can you tell me how these 2 build steps work? Thanks!

File: .github/workflows/build-release.yaml
Comment (abinavpp) [inline .github/workflows/build-release.yaml]: Can we use nproc/sysctl -n hw.ncpu here as well?

File: cmake/templates/BuildInfo.h.in
Comment (abinavpp) [reply cmake/templates/BuildInfo.h.in]: Forgot about the older versions! Whatever that's convenient for you and doesn't break --version output is fine with me :)

File: .github/workflows/build-release.yaml
Comment (antonbaliasnikov) [reply .github/workflows/build-release.yaml:149]: No, unfortunately I can't unify Windows now... for 0.8.x it works without issues with MinGW, but I had to return this ugly solution to have compatibility with older versions... The `action.yml` will be in use for Linux and MacOS, but for Windows we have to be on this boilerplate powershell :(

File: .github/workflows/build-release.yaml
Comment (hedgar2017) [reply .github/workflows/build-release.yaml]: Yes, IIRC we decided to remove this hardcode.

File: cmake/templates/BuildInfo.h.in
Comment (hedgar2017) [reply cmake/templates/BuildInfo.h.in]: Yes, just replace `ZKEVM` with `ZKSYNC` here.

File: .github/workflows/build-release.yaml
Comment (antonbaliasnikov) [reply .github/workflows/build-release.yaml]: To be honest, I'm afraid to touch this powershell part... we can try to use `[Environment]::ProcessorCount` here, but as I don't have access to this Windows host, I'm not 100% sure it will work. I prefer to keep it hardcoded 10 for Windows for now.

File: cmake/templates/BuildInfo.h.in
Comment (hedgar2017) [reply cmake/templates/BuildInfo.h.in]: `zksolc` doesn't rely on the specific words, just on their place in the output.

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
@antonbaliasnikov have you checked if there released binary name matches the format of the latest release?

File: cmake/templates/BuildInfo.h.in
Comment (antonbaliasnikov) [reply cmake/templates/BuildInfo.h.in]: Let's keep it as it is then, thanks for this nit anyway!

File: .github/workflows/build-release.yaml
Comment (hedgar2017) [reply .github/workflows/build-release.yaml]: Ah, if it's only for Windows, it's fine.
What strange Windows machine has 10 threads by the way? 😅

Comment (hedgar2017) [reply .github/workflows/build-release.yaml]: Maybe change it to the actual number of threads of our runner.

Comment (antonbaliasnikov) [reply .github/workflows/build-release.yaml]: I'll ask DevOps to give me access to it, I don't have creds.

**Comment by hedgar2017** [comment]
@antonbaliasnikov I'm granting you admin rights to merge these PRs through those failed tests or even without waiting for the whole workflow to finish.
So @abinavpp can move on to rebasing and fixing the patch.

**Comment by antonbaliasnikov** [comment]
> @antonbaliasnikov I'm granting you admin rights to merge these PRs through those failed tests or even without waiting for the whole workflow to finish. So @abinavpp can move on to rebasing and fixing the patch.

Ok, these failures are caused by the internal compiler error:
`E-M2B3 0.8.25    INVALID tests/solidity/simple/try_catch/nested_2.sol Failed to compile sources: `solc` errors found: ["InternalCompilerError: Assembly exception for bytecode\n\n"]`

File: .github/workflows/build-release.yaml
Comment (antonbaliasnikov) [reply .github/workflows/build-release.yaml]: Ok, so, it's our common runners, we're using `windows-2019-github-hosted-16core`... it contains 16 cores by default, so we can try to put 16.

**Comment by antonbaliasnikov** [comment]
> @antonbaliasnikov have you checked if there released binary name matches the format of the latest release?

Yes, I checked before, but I'll do one more final check just to make sure.

**Comment by abinavpp** [comment]
> > @antonbaliasnikov I'm granting you admin rights to merge these PRs through those failed tests or even without waiting for the whole workflow to finish. So @abinavpp can move on to rebasing and fixing the patch.
> 
> Ok, these failures are caused by the internal compiler error: `E-M2B3 0.8.25 INVALID tests/solidity/simple/try_catch/nested_2.sol Failed to compile sources: `solc` errors found: ["InternalCompilerError: Assembly exception for bytecode\n\n"]`

This will be fixed by #549

**Review by hedgar2017 — APPROVED** [review]

**Review by abinavpp — APPROVED** [review]
Thank you!

**Pull request merged**
